### ISSUE TITLE
Some issues regarding your library

### ISSUE THREAD

**Issue by aytekinar** [stagadish/NNplusplus#1]
Hey,

Today I have got to know about your library thru your post in G+. I am also trying to write a similar library/toolbox to be used in optimization.

At first look at your library, I can say that it looks neat. However, it would be great if you considered using templates in your data structures. Or, if you would like to go for inheritance for some reason, then you can abstract some functionality in your base classes.

What I mean by the above comment is that, for example, you could use an `activation function` abstraction and keep a pointer inside your `NeuralNet` class to support different activation functions, such as:

``` c++
#include <iostream>
#include <vector>

template <typename T>
class IFunction {
public:
  virtual T operator()(T variable) const = 0;
  virtual ~IFunction() = default;
};

template <typename T>
class PReLU : public IFunction<T> {
private:
  T alpha_;
public:
  PReLU(T alpha = T { 0 }) : alpha_ {alpha} { }
  T operator()(T variable) const override {
    T zero { 0 };
    return variable < zero ? alpha_*variable : variable;
  }
};

template <typename T>
class Identity : public IFunction<T> {
public:
  T operator()(T variable) const override {
    return variable;
  }
};

int main (int argc, char* argv[]) {
  Identity    <double>  identity_function {         };
  PReLU       <double>  relu_function     {         };
  PReLU       <double>  absolute_value    { -1.     };

  IFunction   <double>  *generic_func_ptr { nullptr };

  size_t numelems { 100ul };
  double xmin { -5. }, xmax { 5. }, dx {(xmax - xmin)/(numelems-1)};
  std::vector <double>  xvalues ( numelems );
  for (size_t idx = 0; idx < xvalues.size(); idx++)
    xvalues[idx] = xmin + idx*dx;

  // either directly call as if your instances were functions (actually, they
  // are functors right now, with their operator()'s overloaded)
  for (auto x : xvalues)
    std::cout << "x: " << x
              << ", identity_function(x): " << identity_function(x)
              << ", relu_function(x): " << relu_function(x)
              << std::endl;

  // or, use polymorphism
  generic_func_ptr = &identity_function;
  for (auto x : xvalues)
    std::cout << "x: " << x
              << ", generic_func_ptr->operator()(x): "
              << generic_func_ptr->operator()(x)
              << std::endl;

  // or, both
  generic_func_ptr = &relu_function;
  for (auto x : xvalues)
    std::cout << "x: " << x
              << ", (*generic_func_ptr)(x): " << (*generic_func_ptr)(x)
              << std::endl;

  // absolute value function
  for (auto x : xvalues)
    std::cout << "x: " << x
              << ", absolute_value(x): " << absolute_value(x)
              << std::endl;

  return 0;
}
```

I hope this helps :) Good luck with your code!...

**Comment by stagadish** [comment]
Yes, this is indeed a very desirable functionality. Although, the ability to modify the activation function so freely should come hand in hand with the ability to modify its derivation for the gradient decent process during the back propagation step. Another functionality that I would love to add to my class. Eventually.  

As mentioned in my [README](https://github.com/stagadish/NNplusplus/blob/master/README.md#nn), more than anything, this library was (and still is) written for learning purposes. My main goal was to implement a neural net that actually works. That is, LEARN. I never intended it to end up in any professional or commercial use (or even research).  
I would, however, encourage newcomers to machine learning to experiment with my class, especially in C++, since it is so easy to install, read, initialize, and play with.  

There are other, more pressing matters that I would like to address before adding these kinds of abstractions. For example, I need better matrix multiplications. More convenient Matrix constructors for arrays, vectors, and std::initializer_list. I want to add multiple epoch learning with early stopping, etc.

Nevertheless, thank you for your awesome feedback and for taking the time to write out such a descriptive example!

TL;DR - I agree with your comment. I think that it would improve the code significantly and allow for more control if this library was written for professional/commercial uses. But since this is not the case, I think this kind of optimization is lower on my priority list.

**Comment by aytekinar** [comment]
No offense --- I did not mean anything bad. I just wanted to point out to some generalization without much code overhead. You can rely on templates and some basic inheritance, and still learn. Well, you can learn even more in this case. Eventually, when you and/or newcomers are using the source you have provided, they might need to experiment with the functionality by plugging in their favourite loss functions.

Above example can still be of help. Either you can extend the classes with `gradient` methods, or you can redefine the `operator()` method operating on an input variable and returning both the loss/function value and the gradient as its output parameters when either or both are wanted. Anyways, you seem to be knowing what you are supposed to do.

Cheers!

**Comment by stagadish** [comment]
No offense taken! And I hope I didn't come off too defensive. Was not my intention :)
I do agree there's a lot of learning to do from what you proposed, and  like I said, I agree with your point. I am definitely going to refer back to your example when I want to extend user control by adding more hyper-parameters.  

Thanks again for your wonderful input.  

Cheers!
